src/Grpc.AspNetCore.Server/Internal/PercentEncodingHelpers.cs

```diff
@@ -0,0 +1,218 @@
+                    else if (IsUnreservedCharacter(current))
+                    {
+                        span[writePosition++] = current;
+                    }
+                    else
+                    {
+                        EscapeAsciiChar(ref span, ref writePosition, current);
+                    }
+                }
+            }
+        }
+
+        private static void EscapeAsciiChar(ref Span<char> span, ref int writePosition, char current)
+        {
+            span[writePosition++] = '%';
+            span[writePosition++] = HexChars[current >> 4];
+            span[writePosition++] = HexChars[current & 15];
+        }
+
+        private static int GetUnicodeCharCount(string value, int currentIndex)
+        {
+            var unicodeCharCount = 0;
+            var maxSize = Math.Min(value.Length - currentIndex, MaxUnicodeCharsReallocate - 1); // Leave a character for possible low surrogate
+            for (; unicodeCharCount < maxSize && value[currentIndex + unicodeCharCount] > AsciiMaxValue; unicodeCharCount++)
+            {
+            }
+
+            if (char.IsHighSurrogate(value[currentIndex + unicodeCharCount - 1]))
+            {
+                if (unicodeCharCount < value.Length - currentIndex && char.IsLowSurrogate(value[currentIndex + unicodeCharCount]))
+                {
+                    // Last character is a high surrogate so check ahead to see if it is followed by a low surrogate and include
+                    unicodeCharCount++;
+                }
+            }
+
+            return unicodeCharCount;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsUnreservedCharacter(char c)
+        {
+            return ((PercentEncodingUnreservedBitField[c / 8] >> (c % 8)) & 1) != 0;
+        }
+
+        // TODO(JamesNK): Remove WIP decode once we are certain Uri.UnescapeData works correctly
+#if Decode
+        public static string PercentDecode(string value)
+        {
+            var writeLength = 0;
+            var readIndex = 0;
+            while (readIndex < value.Length)
+            {
+                if (value[readIndex] == '%')
+                {
+                    if (readIndex + 2 < value.Length && IsValidHex(value[readIndex + 1]) && IsValidHex(value[readIndex + 2]))
+                    {
+                        readIndex += 3;
+                        writeLength++;
+                    }
+                    else
+                    {
+                        readIndex++;
+                        writeLength++;
+                    }
+                }
+                else
+                {
+                    readIndex++;
+                    writeLength++;
+                }
+            }
+
+            if (writeLength == value.Length)
+            {
+                return value;
+            }
+
+            return string.Create(writeLength, value, Decode);
+
+            static void Decode(Span<char> span, string s)
```

Comment:
The implementation of this method is incorrect. It's decoding individual bytes and widening those bytes to chars when producing the destination string. In effect, that means it's decoding data as ISO-8859-1. This is an asymmetry with the encoding method, which uses UTF-8.

Additionally, per Jan's comment at https://github.com/grpc/grpc-dotnet/pull/223/files#r280808299, if you want to be compliant with the spec you need to include error handling behavior here if you're given invalid input.

I personally believe an error should be returned in this case, even though the spec explicitly calls it out as a disallowed behavior. My opinion is that the current wording of the spec is bound to introduce security vulnerabilities into implementations which are trying to be fully compliant because it encourages such implementations to introduce complexity into code paths which process untrusted input. But if I am alone in this opinion I'm not going to force the argument.

Reply:
All the decode logic here is being replaced with Uri.UnescapeData